#676 - Add _projection method for set - hyperplane lazy intersection

mforets opened the pull request:
Closes #676.

mforets commented:
Apart from the unit test that i'll add now, it is ready from my side.

I gave the options `lazy_linear_map` and `lazy_2d_intersection` such that their default values coincide with the given algorithm in Le Guernic.

schillic commented on src/Intersection.jl:
```diff
@@ -308,6 +308,11 @@ or a hyperplane in a given direction.
                          Golden Section method) 
+    * `"projection"`  -- evaluate the support function by reducing the problem to
+                         the 2D intersection of a rank 2 linear transformation of
+                         the given compat set in the plane generated by the given
```
compact

schillic commented on src/Intersection.jl:
```diff
@@ -308,6 +308,11 @@ or a hyperplane in a given direction.
+                         the 2D intersection of a rank 2 linear transformation of
+                         the given compat set in the plane generated by the given
+                         direction `d` and the hyperplane's normal vector `n`;
+                         only valid for intersection with a hyperplane   
```
Maybe move the last sentence to the top; otherwise the `hyperplane` in the text is confusing.

schillic commented on src/Intersection.jl:
```diff
@@ -475,8 +490,38 @@ end # _line_search
+
+- `ℓ`      -- direction
+- `X`      -- set
+- `H`      -- hyperplane
```
spaces

schillic commented on src/Intersection.jl:
```diff
@@ -475,8 +490,38 @@ end # _line_search
+    _projection(ℓ, X, H::Union{Hyperplane{N}, Line{N}}; kwargs...) where {N}
+
+Given a compact and convex set ``X`` and a hyperplane ``H = \\{x: n ⋅ x = γ \\}``,
+calculate ...
```
"...", now I am excited :smile:

mforets replied:
haha sorry. forgot to write the maths

schillic commented on src/Intersection.jl:
```diff
@@ -475,8 +490,38 @@ end # _line_search
+
+The support function of ``X ∩ H`` along direction ``ℓ``.
+"""
+function _projection(ℓ, X, H::Union{Hyperplane{N}, Line{N}};
```
`ℓ` -> `d`?

mforets replied:
if u don't mind i would keep `ℓ` here so that it matches the original algorithm

schillic commented on src/Intersection.jl:
```diff
@@ -475,8 +490,38 @@ end # _line_search
+                     lazy_2d_intersection=true,
+                     kwargs...) where {N}
+
+    options = Dict(kwargs)
```
`options` is never actually used.

mforets replied:
correct, i have to remove it 👍

schillic commented on src/Intersection.jl:
```diff
@@ -475,8 +490,38 @@ end # _line_search
+    options = Dict(kwargs)
+    n = H.a                  # normal vector to the hyperplane
+    γ = H.b                  # displacement of the hyperplane
+    Πnℓ = vcat(n', ℓ')       # projection map
```
`Πnℓ` - cool idea

schillic commented on src/Intersection.jl:
```diff
@@ -475,8 +490,38 @@ end # _line_search
+
+    Lγ = Line(x_dir, γ)
+
+    Snℓ = lazy_linear_map ? LinearMap(Πnℓ, X) : linear_map(Πnℓ, X)
```
In the spirit of above, what about `\bigcap<TAB>Sℓ` or `S\bigcap<TAB>ℓ` (resp. `d` instead of `ℓ`)?

mforets replied:
nice! we can use the `\bigcap` notation in the hybrid algorithm too 👍

mforets replied:
hmm better if i use `Xnℓ` since the given set is `X`

schillic replied:
What does the notation mean?

mforets commented:
Thanks for the comments, i've addressed them all.

I've also reworked the return statement and added a new keyword argument. The idea is to have more control on how the 2D intersection is performed. I use `nothing` as default type because i want to touch as less as possible the call to `ρ(y_dir, Sℓ⋂Lγ)`.
